src/ExecWin32.cpp

```diff
@@ -485,124 +296,196 @@ ExecResult ExecuteProcess(
+      for (i = 0, len = strlen(out_new_cmd_line); i < len; ++i)
       {
-        if (new_cmd[i] == '\n')
+        if (out_new_cmd_line[i] == '\n')
         {
-          new_cmd[i] = ' ';
+          out_new_cmd_line[i] = ' ';
         }
       }
-
-      result.m_ReturnCode = Win32Spawn(job_id, new_cmd, env_vars, env_count, annotation, echo_cmdline ? cmd_line : NULL, force_use_tty);
-      return result;
     }
   }
-  else
+  return true;
+}
+
+static void CleanupResponseFile(const char* responseFile)
+{
+  if (*responseFile != 0)
+    remove(responseFile);
+}
+
+ExecResult ExecuteProcess(
+  const char*         cmd_line,
+  int                 env_count,
+  const EnvVariable*  env_vars,
+  MemAllocHeap*       heap,
+  int                 job_id,
+  bool                stream_to_stdout)
+{
+  STARTUPINFO sinfo;
+  ZeroMemory(&sinfo, sizeof(STARTUPINFO));
+
+  sinfo.cb = sizeof(STARTUPINFO);
+
+  if (!stream_to_stdout)
   {
-    size_t i, len;
+    sinfo.hStdOutput = GetOrCreateTempFileFor(job_id, Stream::StdOut);
+    sinfo.hStdError = GetOrCreateTempFileFor(job_id, Stream::StdErr);
+    sinfo.hStdInput = GetStdHandle(STD_INPUT_HANDLE);
+    sinfo.dwFlags |= STARTF_USESTDHANDLES;
+  }
 
-    /* Drop any newlines in the command line. They are needed for response
-     * files only to make sure the max length doesn't exceed 128 kb */
-    strncpy(new_cmd, cmd_line, sizeof(new_cmd));
-    new_cmd[sizeof(new_cmd)-1] = '\0';
+  char buffer[8192];
+  char env_block[128 * 1024];
+  WCHAR env_block_wide[128 * 1024];
+  size_t env_block_length = 0;
+  if (!MakeEnvBlock(env_block, sizeof(env_block) - 2, env_vars, env_count, &env_block_length))
+    CroakAbort("env block error; too big?\n");
 
-    for (i = 0, len = strlen(new_cmd); i < len; ++i)
-    {
-      if (new_cmd[i] == '\n')
-      {
-        new_cmd[i] = ' ';
-      }
-    }
+  if (!MultiByteToWideChar(CP_UTF8, 0, env_block, (int)env_block_length, env_block_wide, sizeof(env_block_wide)))
+    CroakAbort("Failed converting environment block to wide char\n");
+
+  ExecResult result;  
+  char new_cmd[8192];
+  char responseFile[MAX_PATH];
+  ZeroMemory(&result, sizeof(ExecResult));
+  ZeroMemory(&responseFile, sizeof(responseFile));
+  ZeroMemory(&new_cmd, sizeof(new_cmd));
 
-    result.m_ReturnCode = Win32Spawn(job_id, new_cmd, env_vars, env_count, annotation, echo_cmdline ? cmd_line : NULL, force_use_tty);
+  if (!SetupResponseFile(cmd_line, new_cmd, sizeof new_cmd, responseFile, sizeof responseFile))
     return result;
+
+  const char* cmd_to_use = new_cmd[0] == 0 ? cmd_line : new_cmd;
+  _snprintf(buffer, sizeof(buffer), "cmd.exe /c \"%s\"", cmd_to_use);
+  buffer[sizeof(buffer) - 1] = '\0';
+
+  HANDLE job_object = CreateJobObject(NULL, NULL);
+  if (!job_object)
+    CroakErrno("ERROR: Couldn't create job object.");
+  
+  PROCESS_INFORMATION pinfo;
+
+  if (!CreateProcessA(NULL, buffer, NULL, NULL, TRUE, CREATE_SUSPENDED | CREATE_UNICODE_ENVIRONMENT, env_block_wide, NULL, &sinfo, &pinfo))
```

Comment:
This should use Unicode version too.

Reply:
env_block_wide is supposed to be terminated by two NULL characters according to documentation (4 bytes since it's wide).